src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -658,17 +659,41 @@ internal void AddNewItemsOfItemType(string itemType, ICollection<ProjectItemInst
             IEnumerable<ProjectItemInstance> itemsToAdd = group;
             if (doNotAddDuplicates)
             {
-                // Remove duplicates from the inputs.
-                itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
 
                 // Ensure we don't also add any that already exist.
                 var existingItems = GetItems(itemType);
+              
+                var existingItemsHashSet = existingItems.ToHashSet(ProjectItemInstance.EqualityComparer);
+
                 if (existingItems.Count > 0)
                 {
-                    itemsToAdd = itemsToAdd.Where(item => !existingItems.Contains(item, ProjectItemInstance.EqualityComparer));
+
+                    var deduplicatedItemsToAdd = new List<ProjectItemInstance>();
+                    foreach (var item in itemsToAdd)
+                    {
+                        if (existingItemsHashSet.Add(item))
+                        {
+                            deduplicatedItemsToAdd.Add(item);
+                        }
+                    }
+                    itemsToAdd = deduplicatedItemsToAdd;
+                }
+                else
+                {
+                    // Remove the duplicates in case we're not concerned with the existing items.
+                    itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
                 }
             }
 
+            if (doNotAddDuplicates)
+            {
+                logFunction?.Invoke(itemsToAdd.ToList());
+            }
+            else {
+                var groupAsList = group as List<ProjectItemInstance>;
+                logFunction?.Invoke(groupAsList ?? group.ToList());  
```

Comment:
The problem with these `?`-heavy expressions is that they're hard to debug. You can't put a breakpoint in one branch to see whether it gets hit. Regular if statements are much better even if they take more lines. They're also straightforward and easy to read.

Reply:
also GitHub doesn't show this, but there are two unnecessary spaces at the end of this line, let's remove them.

Reply:
I thought about this and this particular case of ?? seems fine. 

I'd maybe consider extracting the null check for logFunction to a separate if statement:
`if (logFunction != null) { ... }` this way the groupAsList cast doesn't have to run when the logFunction is null

Reply:
extracted the function.